### ISSUE TITLE
Recorder records lr and mom on_batch_begin 

### ISSUE THREAD

**Issue by kasparlund** [fastai/fastai#1511]
OneCycleScheduler assigns lr and moms to opt in on_batch_end but Record records them in on_batch_begin. The start and end of the recording are therefore wrong

> class Recorder(LearnerCallback):
>     def on_batch_begin(self, train, **kwargs:Any)->None:
>         "Record learning rate and momentum at beginning of batch."
>         if train:
>             self.lrs.append(self.opt.lr)
>             self.moms.append(self.opt.mom)

**assigned OneCycleScheduler.on_batch_end** 
on_batch_end lr:0.0008945512935455861, mom:0.7268941421369995
on_batch_end lr:0.0022254487064544146, mom:0.7731058578630006
on_batch_end lr:0.002863413485248608, mom:0.7952574126822434
on_batch_end lr:0.00298072458933806, mom:0.7993307149075716
on_batch_end lr:0.0029973761725601263, mom:0.7999088948805599
on_batch_end lr:0.0029999888464423605, mom:0.7999996127236931
on_batch_end lr:0.0029999480539389853, mom:0.7999981963173259
on_batch_end lr:0.002999758081361655, mom:0.7999916000472798
on_batch_end lr:0.0029988736294626643, mom:0.7999608899118981
on_batch_end lr:0.002994761513136387, mom:0.7998181080950135
on_batch_end lr:0.002975763404318112, mom:0.7991584515388234
on_batch_end lr:0.0028904905921705096, mom:0.7961975900059205
on_batch_end lr:0.002552238920189802, mom:0.7844527402843682
on_batch_end lr:0.0016705512669004051, mom:0.753838585656264
on_batch_end lr:0.0006967740620810522, mom:0.7200268771555921
on_batch_end lr:0.00026695058078132544, mom:0.7051024507215737
on_batch_end lr:0.0001528689569031804, mom:0.7011412832258048
**on_batch_end lr:0.00012712115469110815, mom:0.7002472623156635**

recorded in Recorder.on_batch_begin:
learn.recorder.lrs:
**[0.003,**
 0.0008945512935455861,
 0.0022254487064544146,
 0.002863413485248608,
 0.00298072458933806,
 0.0029973761725601263,
 0.0029999888464423605,
 0.0029999480539389853,
 0.002999758081361655,
 0.0029988736294626643,
 0.002994761513136387,
 0.002975763404318112,
 0.0028904905921705096,
 0.002552238920189802,
 0.0016705512669004051,
 0.0006967740620810522,
 0.00026695058078132544,
 0.0001528689569031804]

thus missing the value 0.00012712115469110815

**Comment by sgugger** [comment]
It is missing that last value, but that last value was also never used (since it's on the last on_batch_end). I'm not sure why you think there is a bug?

**Comment by kasparlund** [comment]
ok so may not a bug in the last value but certainly in the first. Surely OptimWrapper is not using the max learningrate (0.003) in the first epoch ?
[0.003,
0.0008945512935455861,

**Comment by sgugger** [comment]
Ah yes, the first value recorded should be the max divided by the div_factor. Let me check if it's used by the optimizer or not.

**Comment by sgugger** [comment]
What's your code, I can't reproduce this?

**Comment by kasparlund** [comment]
my mistake i cannot reproduce af latest update on dev
